setuptools/dist.py

```diff
@@ -138,18 +153,22 @@ def invalid_unless_false(dist, attr, value):
     raise DistutilsSetupError(f"{attr} is invalid.")
 
 
-def check_requirements(dist, attr, value):
+@overload
+def check_requirements(dist, attr: str, value: set) -> NoReturn: ...
+@overload
+def check_requirements(dist, attr: str, value: _StrOrIter) -> None: ...
+def check_requirements(dist, attr: str, value: _StrOrIter) -> None:
     """Verify that install_requires is a valid requirements list"""
     try:
         list(_reqs.parse(value))
-        if isinstance(value, (dict, set)):
+        if isinstance(value, set):
```

Comment:
I find it curious that this is the only function in this file checking for a "not set" rather than a "tuple or list". Is there any expectation to allow working with any iterable other than tuple or list? Bbasically the same question I had in https://github.com/pypa/setuptools/pull/4575#discussion_r1720847729

> I also realize that whilst this is all technically type-safe, we may not want to allow `dict` for a user-facing reason: If a user is told they can use a dict here, they're likely going to expect that it means they can use a nested value of some sort, not a "dict with unused values". I also think that forcing a user to coerce their Iterable into a tuple or list in their config file so niche it should be fine.

Reply:
In general, I try not to be overly-prescriptive about input types. For example, accept `Iterable` or `Sequence` so as not to unnecessarily constrain the user. Probably only `tuple` or `list` are used commonly, but I'd be open to allowing users to pass in any ordered object with an `__iter__`.

In this case, we want to assert that whatever value was passed is ordered. The `isinstance(, set)` check is a poor approximation of that, and it's there mainly as a rough check that the user didn't make a silly mistake.